dtschema/schemas/i2c/i2c-controller.yaml

```diff
@@ -44,6 +44,18 @@ properties:
       Number of nanoseconds the SCL signal takes to rise; t(r) in the I2C
       specification.
 
+  i2c-scl-clk-low-timeout-ms:
+    description:
+      Number of miliseconds the clock line needs to be pulled down in order
+      to force a waiting state.
+
+  i2c-scl-has-clk-low-timeout:
+    type: boolean
+    description:
+      Indicates whether the controller implements the feature of wait
+      induction through SCL low, with the timeout being implemented
+      internally by the controller.
+
```

Comment:
Why can't this property be implied by the controller's compatible string? Is it board specific?

Reply:
Hi Rob,

no, this is not board specific, this is i2c standard. In the commit log you will find the exact reference in the standard document. It's a small property of the standard used by only one controller so far. But, with Ryan Chen's patch, now we have already two controllers using it (in a different way, though).

In my opinion it should go here. But please, let me know if you think it should go somewhere else.

Meantime I will fix property/properties.

Thanks,
Andi

Reply:
Hi Krzysztof,

yes, this is board specific for all those boards who are implementing the clock stretching i2c from the standard.

Per se clock stretching is not related to the multi-master property and it's marked as optional even in multi-master controllers. The dependency between clock-stretching and multi-master should be specific to the hardware that Chen is implementing.

Andi

Reply:
For example clock stretching can also happen when the master wants to reset the communication because something wrong happened. But it's an optional property that apparently almost no one is using or implementing in upstream drivers.

Reply:
> Per se clock stretching is not related to the multi-master property and it's marked as optional even in multi-master controllers. The dependency between clock-stretching and multi-master should be specific to the hardware that Chen is implementing.

I never got answer also from Ryan Chen why someone would not enable timeout for a multi-master setup. I understand it is optional, but why one would not enable it?

> For example clock stretching can also happen when the master wants to reset the communication because something wrong happened.

How lack of this timeout property stops the device from using clock stretching? Your commit msg said:

> This property specifies whether the I2C controller implements the clock stretching property.

... so compatible says that. The properties do not say what is implemented by controller, because otherwise you would start documenting entire I2C spec in the bindings. The compatible says it all.

> But it's an optional property that apparently almost no one is using or implementing in upstream drivers.

Driver do not matter for bindings in terms of argument whether property is valid or not.

Reply:
> > Per se clock stretching is not related to the multi-master property and it's marked as optional even in multi-master controllers. The dependency between clock-stretching and multi-master should be specific to the hardware that Chen is implementing.
> 
> I never got answer also from Ryan Chen why someone would not enable timeout for a multi-master setup. I understand it is optional, but why one would not enable it?

Maybe it's not optional in his hardware. For what I understood his hardware has implemented the multimaster depending on the clock stretching. Which is weird, but he knows best as he has the datasheets.

> > For example clock stretching can also happen when the master wants to reset the communication because something wrong happened.
> 
> How lack of this timeout property stops the device from using clock stretching? Your commit msg said:

"timeout" and "clock stretching" are synonyms. Apparently people prefer to call it timeout, so does the SMBUS specification. The i2c specification calls it clock stretching. I use them both. Is it confusing?

> > This property specifies whether the I2C controller implements the clock stretching property.
> 
> ... so compatible says that. The properties do not say what is implemented by controller, because otherwise you would start documenting entire I2C spec in the bindings. The compatible says it all.

It does, maybe I haven't been clear enough. We have these two properties:

- i2c-scl-clk-low-timeout-us: where the clock stretching (or timeout) is implemented in the driver (see i2c-mpc), or think of i2c-gpio where everything is done in software.
- i2c-scl-has-clk-low-timeout: it's implemented by the controller (Ryan's case) where you ask the controller to start the clock stretching until it ends. (After that there is the master switch, _in Ryan's case_)

Is it clear?

> > But it's an optional property that apparently almost no one is using or implementing in upstream drivers.
> 
> Drivers do not matter for bindings in terms of argument whether property is valid or not.

I'm not understanding. The first property (-timeout-us) specifies how long should the wait be. Please consider that also i2c slaves might be dependent on this property. Depending on the i2c slave you might need to fine tune the microseconds (milliseconds, indeed :) ) you want to wait. Some slaves might need longer time to have a wait triggered some need a shorter time. So that it can be that the board does not have it, but because of some slaves we might need to implement it in software. (Is it the case of i2c-mpc?)

In the case of Ryan's hardware apparently the board does have it in hardware, but as an optional feature. Apparently needed for the multi-master (which is an optional configuration itself).

If, for example, you have slaves that do not understand what clock stretching is (like most of the sensors) then you don't need it.

Reply:
> > I never got answer also from Ryan Chen why someone would not enable timeout for a multi-master setup. I understand it is optional, but why one would not enable it?
> 
> Maybe it's not optional in his hardware.

If it was not optional in his hardware, you do not need that property. Why would you add to the DT bindings a boolean property which is always required? Just imply it from compatible and we are done.

> For what I understood his hardware has implemented the multimaster depending on the clock stretching. Which is weird, but he knows best as he has the datasheets.

If his multi-master setup depends on it, then he does not need that property.

> > How lack of this timeout property stops the device from using clock stretching? Your commit msg said:
> 
> "timeout" and "clock stretching" are synonyms. Apparently people prefer to call it timeout, so does the SMBUS specification. The i2c specification calls it clock stretching. I use them both. Is it confusing?

No, but I asked how lack of this property stops the device (controller) from using it. We have lengthy discussion why do you need that property and still no answer.

> > > This property specifies whether the I2C controller implements the clock stretching property.
> > 
> > 
> > ... so compatible says that. The properties do not say what is implemented by controller, because otherwise you would start documenting entire I2C spec in the bindings. The compatible says it all.
> 
> It does, maybe I haven't been clear enough. 

If compatible says, there is no need for this property.

> We have these two properties:
> 
> * i2c-scl-clk-low-timeout-us: where the clock stretching (or timeout) is implemented in the driver (see i2c-mpc), or think of i2c-gpio where everything is done in software.
> * i2c-scl-has-clk-low-timeout: it's implemented by the controller (Ryan's case) where you ask the controller to start the clock stretching until it ends. (After that there is the master switch, _in Ryan's case_)
> 
> Is it clear?

This yes, but it's not what exactly I was asking about.

> 
> > > But it's an optional property that apparently almost no one is using or implementing in upstream drivers.
> > 
> > 
> > Drivers do not matter for bindings in terms of argument whether property is valid or not.
> 
> I'm not understanding. 

Which part? That driver does not matter? DTS describes hardware, not driver. Devicetree describes DTS thus it still describes hardware, not driver. Using driver as an argument therefore usually does not apply, because driver does not matter.

> The first property (-timeout-us) specifies how long should the wait be. Please consider that also i2c slaves might be dependent on this property. 

That's the key point and it was nowhere mentioned up to here.

> Depending on the i2c slave you might need to fine tune the microseconds (milliseconds, indeed :) ) you want to wait. Some slaves might need longer time to have a wait triggered some need a shorter time. So that it can be that the board does not have it, but because of some slaves we might need to implement it in software. (Is it the case of i2c-mpc?)

So this is the rationale behind the timeout property with value. Now the question why do we ever need the bool timeout property.

> 
> In the case of Ryan's hardware apparently the board does have it in hardware, but as an optional feature. Apparently needed for the multi-master (which is an optional configuration itself).
> 
> If, for example, you have slaves that do not understand what clock stretching is (like most of the sensors) then you don't need it.

So is the case here for "i2c-scl-has-clk-low-timeout" property?

Reply:
[Removing the part above, as I think it all comes to this:]

> So is the case here for "i2c-scl-has-clk-low-timeout" property?

OK... I think your logic is:

multi-master is optional -> multimaster depends on timeout -> why have timeout optional?

Because, if I'm reading correctly his driver, he is actually implementing it for other use cases, as well. Besides, he has implemented it both in master and slave side. Check for example the function ast2600_i2c_recover_bus(), that is not related to the multi-master capability, but uses this timeout for bus recovery in case of error. The clock stretching was initially made for bus recovery, indeed.

Consider also that in multimaster configurations, masters become slaves and slaves become masters. So that you have a perfect example where you have slaves needing this. "yes, sure, but if we have multimaster defined, we don't need this"... read above.

Be aware that this is a completely optional feature. If in a phone you don't have slaves that understand clock stratching, you leave it out and save yourself some wait_for_completion_timeout(). If you have it (e.g. in complex MCU systems, or cameras), then you have the option to use this. And it's a pretty cool picture, because it's an efficient way to suddenly interrupt a communication (think of i2c's with over 1GHz of bandwidth).

Reply:
Thank you a lot, Krzysztof. I appreciate your review. I will add a better commit log.